network/src/fuzzing.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    protocols::{
+        identity::exchange_handshake,
+        wire::handshake::v1::{HandshakeMsg, MessagingProtocolVersion, SupportedProtocols},
+    },
+    testutils::fake_socket::ReadOnlyTestSocketVec,
+};
+use futures::executor::block_on;
+use libra_config::network_id::NetworkId;
+use libra_types::chain_id::ChainId;
+use proptest::{collection::btree_map, prelude::*};
+
+//
+// Handshake Protocol Fuzzer
+// =========================
+//
+
+/// Serializes a HandshakeMsg by simulating sending it over a socket
+fn serialize_handshake_message(handshake_msg: &HandshakeMsg) -> Vec<u8> {
+    // serialize with LCS
+    let handshake_msg = lcs::to_bytes(handshake_msg).unwrap();
+    // prepend a 2-byte prefix indicating the message length
+    let mut serialized = (handshake_msg.len() as u16).to_be_bytes().to_vec();
+    serialized.extend_from_slice(&handshake_msg);
+    serialized
+}
+
+/// Fuzzing the handshake protocol, which negotiates protocols supported by both
+/// the client and the server.
+/// At the moment, fuzzing the client or the server leads to the same logic.
+pub fn fuzz_network_handshake_protocol_exchange(self_handshake: &HandshakeMsg, data: Vec<u8>) {
+    // fake socket to read the other peer's serialized HandshakeMsg from
+    let mut fake_socket = ReadOnlyTestSocketVec::new(data);
+    fake_socket.set_trailing();
+
+    // fuzz the network exchange of HandshakeMsg first
+    let _ = block_on(async move {
+        if let Ok(remote_handshake_msg) = exchange_handshake(self_handshake, &mut fake_socket).await
+        {
+            // then perform the negotiation
+            let _ = self_handshake.perform_handshake(&remote_handshake_msg);
```

Comment:
I really wish there was a better way in rust to throw away the return values.  I guess we don't really care what the response is?

Reply:
we could in theory test some properties here, for example that the intersection only includes things from both handshake message, but that might be a bit too much :p